pkg/kubelet/kubelet.go

```diff
@@ -1699,8 +1686,9 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
 		return err
 	}
 
-	// Volume manager will not mount volumes for terminated pods
-	if !kl.podIsTerminated(pod) {
+	// Volume manager will not mount volumes for terminating pods
+	// TODO: once context cancellation is added this check can be removed
+	if !kl.podWorkers.IsPodTerminationRequested(pod.UID) {
```

Comment:
We are in a sync* method here, which means we are running in the pod worker and we know for certain the pod is known to the kubelet.  This method is checking whether in the time since syncPod has started, has someone asked us to stop.  That is the equivalent to the previous check 

> for someone running in the pod worker, should this pod keep attempting to create containers?

The `ShouldPodContainersBeTerminating` method is different.  It is running *outside* a pod worker, and can't return true until `syncPod` exits.  So the syncPod gets to know early (because it's the one creating containers) and all other workers have to wait until syncPod exits.

This is the core race the PR addresses - that other loops need to know FOR SURE that no new containers can be started, which only pod worker knows.  The pod worker itself needs to know when syncPod has been requested to stop early, but syncPod at this line of code also knows that containers could still get started in the next loop.  In the future, when context cancellation happens, instead of continuing on to sync pod, we would exit earlier from syncPod, which means that the volume manager sees ShouldPodContainersBeTerminating earlier.  At that point we would remove IsPodTerminationRequested.

I can change this to simply be an exit early, or I can have it check for context cancellation, but that's bring more scope in.  The correct version of this code looks like:

```
func syncPod(ctx context.Context) { 
   ...
   if err := kl.volumeManager.WaitForAttachAndMount(ctx,` pod); err != nil {
     if err == context.Canceled {
       return err
     }
     ... // existing debug logic
  }
  result := kl.containerRuntime.SyncPod(ctx, ...)
  if result is cancelled {
    return context.Canceled
  }
  ... // err checking
  return nil
}
```

There is no deadlock possible that I'm aware of because SyncPod will fail when mounts aren't ready at container create or container start (which is why we wait here in normal paths).